tool/src/org/antlr/v4/automata/ParserATNFactory.java

```diff
@@ -472,24 +472,26 @@ public Handle elemList(List<Handle> els) {
             if ( el.left.getNumberOfTransitions()==1 ) tr = el.left.transition(0);
             boolean isRuleTrans = tr instanceof RuleTransition;
             if ( el.left.getStateType() == ATNState.BASIC &&
+				el.right != null &&
 				el.right.getStateType()== ATNState.BASIC &&
-				tr!=null && (isRuleTrans && ((RuleTransition)tr).followState == el.right || tr.target == el.right) )
-			{
+				tr!=null && (isRuleTrans && ((RuleTransition)tr).followState == el.right || tr.target == el.right) ) {
 				// we can avoid epsilon edge to next el
-				if ( isRuleTrans ) ((RuleTransition)tr).followState = els.get(i+1).left;
-                else tr.target = els.get(i+1).left;
+				Handle handle = i + 1 < els.size() ? els.get(i + 1) : null;
+				if (handle != null) {
+					if (isRuleTrans)
+						((RuleTransition) tr).followState = handle.left;
+					else
+						tr.target = handle.left;
+				}
 				atn.removeState(el.right); // we skipped over this state
 			}
 			else { // need epsilon if previous block's right end node is complicated
 				epsilon(el.right, els.get(i+1).left);
 			}
 		}
 		Handle first = els.get(0);
-		Handle last = els.get(n -1);
-		if ( first==null || last==null ) {
-			g.tool.errMgr.toolError(ErrorType.INTERNAL_ERROR, "element list has first|last == null");
-		}
-		return new Handle(first.left, last.right);
+		Handle last = els.get(n - 1);
```

Comment:
That's a significant change in behavior, that might break users code

Reply:
I've just removed `first==null || last==null` check, `null` elements are being processed later. How it could affect user code? All tests are passing.

Reply:
It no longer throws a tool error, that's a significant change.
Not sure how to reproduce the throwing scenario, but it would be better to have a test for that (which throws with the existing code) and check that it still does with the  new code.
We never know what developers do with development tools, they might be sending some random grammars and expecting an error if the grammar isn't valid...

Reply:
I'm not saying this is important, but not sure the first message can't be improved ? How about putting the undefined rules in quotes ?

Reply:
Maybe it can, but I didn't change it. This is the existing error that is being thrown in https://github.com/antlr/antlr4/blob/dev/tool/src/org/antlr/v4/automata/ParserATNFactory.java#L297 Also, it looks like we don't use quotes for other similar errors, see for example https://github.com/antlr/antlr4/blob/e19c0abea36409dbd1b89035a37f8145b517406d/tool/src/org/antlr/v4/tool/ErrorType.java#L204